### PR TITLE
Add recommendations tracking

### PR DESCRIPTION
**Pull request by mocca102**
- Track recommendation click 
- Track recommendation view

### CODE REVIEW

File: src/components/Autocomplete/SectionItemsList/SectionItemsList.tsx
Comment (mocca102) [inline src/components/Autocomplete/SectionItemsList/SectionItemsList.tsx:19]: Prop Getter for section props that has cnstrc data attributes

File: src/hooks/useCioAutocomplete.ts
Comment (mocca102) [inline src/hooks/useCioAutocomplete.ts:9]: Removed useItems and added it to utils as a function called getItems instead since there was no need for it to be a hook in the first place

File: src/hooks/useCioAutocomplete.ts
Comment (mocca102) [inline src/hooks/useCioAutocomplete.ts:169]: Attaching ref to sections to observe them using useObserveRecommendations hook

File: src/hooks/useDownShift.ts
Comment (mocca102) [inline src/hooks/useDownShift.ts:30]: I believe `!selectedItem?.data?.url` was to check if not product. But I believe it's more accurate to send search submit only if section is Search Suggestion

File: src/hooks/useSections.ts
Comment (mocca102) [inline src/hooks/useSections.ts:23]: Generate refs that will be attached to sections

File: src/hooks/useSections.ts
Comment (mocca102) [inline src/hooks/useSections.ts:49]: ActiveSection equals this expression `zeroStateActiveSections ? zeroStateSections : sections`  that's already defined in the default state and it doesn't change 

Already if zeroStateActiveSections or sections change it gets updated

File: src/hooks/useSections.ts
Comment (mocca102) [inline src/hooks/useSections.ts:54]: ActiveSectionsWithData is reactive and get updated based on what get returned form fetchResults

File: src/hooks/useRecommendationsObserver.ts
Comment (mocca102) [inline src/hooks/useRecommendationsObserver.ts:17]: Enes says we should filter by recommendation sections here

File: src/hooks/useSections.ts
Comment (mocca102) [reply src/hooks/useSections.ts:29]: Do we need this to be a ref?

Comment (esezen) [reply src/hooks/useSections.ts:29]: @mocca102 did you decide this needs to be a ref?

**Review by esezen — APPROVED** [review]
This is looking great! Thanks for making the changes.

I left some small comments, let me know what you think

Lets also get a review from the Professor @deomsj

File: src/utils.ts
Comment (deomsj) [reply src/utils.ts:205]: @esezen and @mocca102,

please see Alex S's perspective on this topic in [this slack thread from 3 months ago](https://constructor.slack.com/archives/C05LF8MTYG6/p1695227539938669?thread_ts=1694453832.486309&cid=C05LF8MTYG6): 

TLDR:
"3. how should num_results_viewed be determined?
Ideally, it would be nice to track all items in the pod (even the ones not visible at first) & corresponding num_results_viewed. Tracking just items that are visible (& corresponding num_results_viewed) is fine as well"

1. Do y'all have reason to believe that this directive has changed? 
2. Given this information, what do you think our next steps are here?

File: src/hooks/useCioAutocomplete.ts
Comment (deomsj) [inline src/hooks/useCioAutocomplete.ts:60]: @mocca102, are we memoizing `activeSectionsWithData`?

in order for this dependency array and useMemo hook to  behave as intended, we'll want to make sure that we are not just passing a reference to an object or array that will change on every re-render. 

dependency arrays should only be passed:
 - primitives
 - memoized objs/arrays
 - objs/arrays that are use the same reference across renders (e.g. variables managed by useState hook)

for reference, more on this topic [here](https://react.dev/reference/react/useMemo#memoizing-a-dependency-of-another-hook)

File: src/hooks/useCioAutocomplete.ts
Comment (deomsj) [inline src/hooks/useCioAutocomplete.ts:185]: @mocca102, can we move this key value pair to the object initially assigned to `attributes` above, or is only relevant to recommendations sections?
`'data-cnstrc-section': section.data[0]?.section,`

Would also refactor to include rec specific attributes inside of the condition and logical branching that explains these differences. How do we feel about something like this:

```javascript
      if (section.type === 'recommendations') {
         attributes['data-cnstrc-recommendations'] = true;
         attributes['data-cnstrc-recommendations-pod-id'] = section.identifier;
      }
```

File: src/hooks/useDownShift.ts
Comment (deomsj) [inline src/hooks/useDownShift.ts:51]: what sections do we want to send this for? 
 - search suggestions only? 
 - products and SS? 
 - content items? 
 - custom sections?
 
 I'm not sure what the answer is here, do either of y'all know, @mocca102 / @esezen ?
 
when we have clear direction on the question above, would love to refactor the logical branching in these blocks of code to make this crystal clear to anyone referencing or building on top of this code

File: src/hooks/useRecommendationsObserver.ts
Comment (deomsj) [inline src/hooks/useRecommendationsObserver.ts:22]: I'm confused. What do we mean by "intersection" here?

File: src/hooks/useRecommendationsObserver.ts
Comment (deomsj) [inline src/hooks/useRecommendationsObserver.ts:28]: are we only passing `trackRecommendationView` as our `callback`? 
 - if so, I'd prefer that we remove this comment and update the argument name to `trackRecommendationView` to make this explicit, then remove this comment.
 - else, let's remove this comment or update it to explain other use cases

File: src/hooks/useRecommendationsObserver.ts
Comment (deomsj) [inline src/hooks/useRecommendationsObserver.ts:34]: what is going on here? can we declare this obj with `root` and `threshold` above and reference it here with some sort of meaningful variable name that makes it easier to understand what is happening here?

File: src/hooks/useRecommendationsObserver.ts
Comment (deomsj) [reply src/hooks/useRecommendationsObserver.ts:22]: whoa... https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API

this is very coooool :)

File: src/utils.ts
Comment (esezen) [reply src/utils.ts:205]: I don't understand why we have `num_results_viewed` along with `result_count` if this is the case. We have been sending the visible count for `num_results_viewed` in the beacons. If Alex thinks `num_results_viewed` should be all, that's perfectly fine by me but I don't understand why we have 2 params. In terms of next steps, I think we are good, the current implementation sends all so we can leave it as is

File: src/hooks/useRecommendationsObserver.ts
Comment (deomsj) [inline src/hooks/useRecommendationsObserver.ts:6]: can we add some brief comment above this function definition describing the scope of responsibility for this hook? We are using some very cool patterns in here that will not be obvious to many developers. I'd love to see explicit documentation for what this hook is doing for us and what needs to be configured elsewhere in the lib code or in a lib consumer's src code in order for this to behave as it is intended

File: src/hooks/useSections.ts
Comment (deomsj) [inline src/hooks/useSections.ts:27]: I may be missing something, but this looks like derived state at this point. 
Can we simplify? 

all of this:
```javascript
  // Define All Sections
  const [activeSections, setActiveSections] = useState<UserDefinedSection[]>(
    zeroStateActiveSections ? zeroStateSections : sections
  );

  useEffect(() => {
    setActiveSections(zeroStateActiveSections ? zeroStateSections : sections);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [zeroStateActiveSections]);
```

to just, this:
this:
```javascript
  const activeSections = zeroStateActiveSections ? zeroStateSections : sections;
```

File: src/hooks/useSections.ts
Comment (deomsj) [inline src/hooks/useSections.ts:65]: why do we need this effect? I'm wondering why we need to manage `activeSections` as its own state here at all instead of deriving it from other state passed in to this hook...

**Review by deomsj — CHANGES_REQUESTED** [review]
nice work on this @mocca102 and @esezen!!! 

I've left a number of ideas, questions, and suggestions as inline comments in this review. Looking forward to hearing your thoughts on these points and having another look at this together on a pairing call or async for another round of review.

I appreciate all of your great work on this & I'll follow your lead on how I can best support with next steps :)

File: src/hooks/useDownShift.ts
Comment (esezen) [reply src/hooks/useDownShift.ts:51]: I believe all autocomplete sections including SS, Products, and custom sections (not to be confused with "custom" sections of this library, I'm talking about index sections)

File: src/hooks/useCioAutocomplete.ts
Comment (mocca102) [reply src/hooks/useCioAutocomplete.ts:60]: I believe we meet the third criteria here 😃 

> objs/arrays that are use the same reference across renders (e.g. variables managed by useState hook)

`activeSectionsWithData` is returned from `useSections` hook and inside it `activeSectionsWithData` is a variable managed by `useState` hook

File: src/hooks/useSections.ts
Comment (mocca102) [reply src/hooks/useSections.ts:27]: I thought I wanted to make it reactive to only zeroStateActiveSections so it updates the state only if zeroStateActiveSections changes. But whenever zeroStateActiveSections, zeroStateSections and sections also change and they're passed as props to the hook so the whole hook rerun every time any way.

TLDR; I thought this was an optimization but I don't think it is anymore

File: src/utils.ts
Comment (mocca102) [reply src/utils.ts:198]: Based on the defined types It should be an object and at this step it should always be defined. So if it's not an object or if it's undefined this means that there's a bug that needs to be addressed. Based on the previous I believe it's better not to add optional chaining since we would need this to throw an error if it's undefined or not an object so we can catch it

File: src/hooks/useCioAutocomplete.ts
Comment (mocca102) [reply src/hooks/useCioAutocomplete.ts:185]: Done https://github.com/Constructor-io/constructorio-ui-autocomplete/pull/107/commits/55c168e54f3bc1568827cba615592e193f2fc3c1#diff-7f1c987b478566cc9cf82a2afc2ee7a7a150ec6047a7801e0ac01c410f7ce453R183

File: src/hooks/useDownShift.ts
Comment (mocca102) [reply src/hooks/useDownShift.ts:53]: The previous logic did fire selects for user custom defined sections. Fixed here https://github.com/Constructor-io/constructorio-ui-autocomplete/pull/107/commits/55c168e54f3bc1568827cba615592e193f2fc3c1#diff-2df40d7bb6359f01b937f7c0c6a7cfbd72521a04f72477d270a0fb94073d430eR52

File: src/hooks/useRecommendationsObserver.ts
Comment (mocca102) [reply src/hooks/useRecommendationsObserver.ts:18]: Let me know what you think https://github.com/Constructor-io/constructorio-ui-autocomplete/pull/107/commits/55c168e54f3bc1568827cba615592e193f2fc3c1#diff-54d8f2b4755e1b9a3357013b05b24a47635a9fbae6dc1ab0116aa0dcaf69aa05R7

File: src/hooks/useRecommendationsObserver.ts
Comment (mocca102) [reply src/hooks/useRecommendationsObserver.ts]: https://github.com/Constructor-io/constructorio-ui-autocomplete/pull/107/commits/55c168e54f3bc1568827cba615592e193f2fc3c1#diff-54d8f2b4755e1b9a3357013b05b24a47635a9fbae6dc1ab0116aa0dcaf69aa05R44

File: src/hooks/useRecommendationsObserver.ts
Comment (mocca102) [reply src/hooks/useRecommendationsObserver.ts]: https://github.com/Constructor-io/constructorio-ui-autocomplete/pull/107/commits/55c168e54f3bc1568827cba615592e193f2fc3c1#diff-54d8f2b4755e1b9a3357013b05b24a47635a9fbae6dc1ab0116aa0dcaf69aa05R34

File: src/hooks/useSections.ts
Comment (mocca102) [reply src/hooks/useSections.ts]: https://github.com/Constructor-io/constructorio-ui-autocomplete/pull/107/commits/55c168e54f3bc1568827cba615592e193f2fc3c1#diff-c72a6a39c2792562569f2b7e56003020f1eefbc541ba6221750d044635f0d1b1R20

File: src/hooks/useSections.ts
Comment (mocca102) [reply src/hooks/useSections.ts]: Removed

File: src/utils.ts
Comment (mocca102) [reply src/utils.ts]: Let me know if this clears things up https://github.com/Constructor-io/constructorio-ui-autocomplete/pull/107/commits/55c168e54f3bc1568827cba615592e193f2fc3c1#diff-39b2554fd18da165b59a6351b1aafff3714e2a80c1435f2de9706355b4d32351R173

File: src/utils.ts
Comment (mocca102) [reply src/utils.ts]: Removed https://github.com/Constructor-io/constructorio-ui-autocomplete/pull/107/commits/55c168e54f3bc1568827cba615592e193f2fc3c1#diff-39b2554fd18da165b59a6351b1aafff3714e2a80c1435f2de9706355b4d32351R173

File: src/utils.ts
Comment (mocca102) [reply src/utils.ts:205]: left as is

File: src/hooks/useRecommendationsObserver.ts
Comment (deomsj) [reply src/hooks/useRecommendationsObserver.ts:18]: YES. Well done sir. I found these explanations to be very, very helpful :)

File: src/hooks/useRecommendationsObserver.ts
Comment (deomsj) [inline src/hooks/useRecommendationsObserver.ts:10]: vocab alignment:
 - "user" - end user who's browser this production code is rendered on
 - "lib consumer" - developer / team who is consuming this library and building on top of it

File: src/utils.ts
Comment (deomsj) [inline src/utils.ts:173]: Four questions:
1. how will the lib consumer know that they can pass these refs? 
2. how will lib consumer know how to pass this ref? 
3. what happens if they don't pass this ref and are using the lib's component interface? 
4. what happens if they don't pass this ref and are using the lib's hook interface?

**Review by deomsj — CHANGES_REQUESTED** [review]
Nicely done @mocca102!

I haven't tried breaking functionality locally yet, but I'm lovin the changes made here from a code clarity / maintainability perspective.

I left a couple more questions & ideas. Holla for pairing sesh or for another review as you like :)

File: src/utils.ts
Comment (esezen) [reply src/utils.ts:205]: I think we can leave the num results viewed as is but can we still add `resultCount` and `resultId`?

Comment (esezen) [reply src/utils.ts:205]: Update here:
https://constructor.slack.com/archives/C05LF8MTYG6/p1702412492166239?thread_ts=1694453832.486309&cid=C05LF8MTYG6
If we finalize this PR before we hear back, we can merge and follow up in a different story

File: src/utils.ts
Comment (mocca102) [reply src/utils.ts:173]: 1. It's a prop on type `SectionConfiguration` https://github.com/Constructor-io/constructorio-ui-autocomplete/pull/107/files/55c168e54f3bc1568827cba615592e193f2fc3c1#diff-c54113cf61ec99691748a3890bfbeb00e10efb3f0a76f03a0fd9ec49072e410aR83
2. same as 1 but we can add more documentation explaining that
3. and 4. There's no change between hooks and component interface as long as they use getSectionProps

We need to add docs for getSectionProps

Comment (mocca102) [reply src/utils.ts]: This is a great callout. This would have made things more clear if added from the beginning. Added here
https://github.com/Constructor-io/constructorio-ui-autocomplete/pull/107/commits/06ced4f6e422f6213c02d08570acb4abba0f084c#diff-8fa4b52909f895e8cda060d2035234e0a42ca2c7d3f8f8de1b35a056537bf199R40 @deomsj

**Review by deomsj — APPROVED** [review]
Findings from dev build on my machine:
 - all storybook tests are passing!
 - all rec view and rec click events look great! (screenshot below for more detail)
 - tried a handfull of stories without recs and both UX and network traffic continue to behave as expected!
 
 Excellent work on this, @mocca102!!!

Screenshot 2023-12-13 at 9 25 47 AM

**Pull request merged**
